src/gt4py/next/program_processors/runners/dace/workflow/translation.py

```diff
@@ -113,69 +113,125 @@ def _has_gpu_schedule(sdfg: dace.SDFG) -> bool:
+        entry_state, "gt_timer_begin", condition=f"{metrics_level} >= {metrics.PERFORMANCE}"
+    )
 
-    # NOTE: We should actually wrap the `StreamSynchronize` function inside a
-    #   `DACE_GPU_CHECK()` macro. However, this only works in GPU context, but
-    #   here we are in CPU context. Thus we can not do it.
-    dace_gpu_backend = dace.Config.get("compiler.cuda.backend")
-    assert dace_gpu_backend in ["cuda", "hip"], f"GPU backend '{dace_gpu_backend}' is unknown."
-    sync_state.add_tasklet(
-        "sync_tlet",
-        inputs=set(),
-        outputs=set(),
-        code=f"{dace_gpu_backend}StreamSynchronize({dace_gpu_backend}StreamDefault);",
+    for source_state in sdfg.source_nodes():
+        if source_state is entry_state:
+            continue
+        sdfg.add_edge(
+            entry_state,
+            source_state,
+            dace.InterstateEdge(condition=f"{metrics_level} < {metrics.PERFORMANCE}"),
+        )
+        sdfg.add_edge(begin_state, source_state, dace.InterstateEdge())
+    assert sdfg.out_degree(begin_state) > 0
+    assert sdfg.out_degree(entry_state) > 1
+
+    tlet_start_timer = begin_state.add_tasklet(
+        "gt_start_timer",
+        inputs={},
+        outputs={"time"},
+        code="""\
+time = static_cast<double>(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+        std::chrono::high_resolution_clock::now().time_since_epoch()).count()
+) / 1e9;
+        """,
         language=dace.dtypes.Language.CPP,
-        side_effects=True,
+    )
+    begin_state.add_edge(
+        tlet_start_timer,
+        "time",
+        begin_state.add_access(start_time),
+        None,
+        dace.Memlet(f"{start_time}[0]"),
     )
 
-    # DaCe [still generates a stream](https://github.com/spcl/dace/blob/54c935cfe74a52c5107dc91680e6201ddbf86821/dace/codegen/targets/cuda.py#L467)
-    #  despite not using it. Thus to be absolutely sure, we will not set that stream
-    #  to the default stream.
-    sdfg.append_init_code(
-        f"__dace_gpu_set_all_streams(__state, {dace_gpu_backend}StreamDefault);",
-        location="cuda",
+    #### 3. Collect the SDFG end timestamp and produce the compute metric.
+    end_state = sdfg.add_state("gt_timer_end")
+    for sink_state in sdfg.sink_nodes():
+        if sink_state is end_state:
+            continue
+        sdfg.add_edge(
+            sink_state,
+            end_state,
+            dace.InterstateEdge(condition=f"{metrics_level} >= {metrics.PERFORMANCE}"),
+        )
+    assert sdfg.in_degree(end_state) > 0
+
+    tlet_stop_timer = end_state.add_tasklet(
+        "gt_stop_timer",
+        inputs={"run_cpp_start_time"},
+        outputs={"duration"},
+        code=sync_code
+        + """
+double run_cpp_end_time = static_cast<double>(
+    std::chrono::duration_cast<std::chrono::nanoseconds>(
+        std::chrono::high_resolution_clock::now().time_since_epoch()).count()
+) / 1e9;
+duration = run_cpp_end_time - run_cpp_start_time;
```

Comment:
for my understanding: in the generated code `duration` is a parameter of some c++ function because it's in `outputs`? Similarly, the `time` from line 184 is output and gets passed as input to this piece?